### PR TITLE
KAZOO-3503: Adding new api port_requests/find/{number}

### PR DESCRIPTION
**Pull request by macpie**

### CODE REVIEW

**Comment by jamesaimonetti** [comment]
@macpie What level of user can access this API? Seems unwise to allow any authenticated user access to the full listing of a port request. Shouldn't it check that the account of the port request is the auth-account or a descendant?

Also, `find` in a URI seems un-rest-like; we try to avoid putting verbs in there. I would say `GET /port_requests?by_number={NUMBER}` since you're still getting a listing. Then you can toggle which view to query based on this query string parameter's existence.

Thoughts?

**Comment by macpie** [comment]
1) There is no access restriction on the others + you would have to know the number but I am not close to the idea tho.

2) I agree with you, it was just the ticket requirement, what should I do there?

**Comment by jamesaimonetti** [comment]
1. `/port_requests` is restricted to the account on the URI. Not sure what you mean by no access restrictions? You can't lookup arbitrary port requests in the system.
2. I read the ticket's URI as an idea, not a hard and fast rule. When this is merged, let the UI know (and update the docs) how to access this URI.

**Comment by macpie** [comment]
1) Seems good enough for me.
2) Ok, I'll do that.

File: applications/crossbar/src/modules/cb_port_requests.erl [hunk 1]
```diff
@@ -715,6 +721,22 @@ summary(Context) ->
...
 %%--------------------------------------------------------------------
+-spec by_number(cb_context:context(), ne_binary()) -> cb_context:context().
+by_number(Context, Number) ->
+    E164 = wnm_util:to_e164(Number),
+    ViewOptions = [{'key', E164}, 'include_docs'],
+    crossbar_doc:load_view(
+        ?ALL_PORT_REQ_NUMBERS
+        ,ViewOptions
+        ,cb_context:set_account_db(Context, ?KZ_PORT_REQUESTS_DB)
+        ,fun normalize_view_results/2
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: `fun(JObj, Acc) -> normalize_view_results(JObj, Acc, Context) end`

File: applications/crossbar/src/modules/cb_port_requests.erl [hunk 2]
```diff
@@ -715,6 +721,23 @@ summary(Context) ->
 %% @doc
 %% @end
 %%--------------------------------------------------------------------
+-spec by_number(cb_context:context(), ne_binary()) -> cb_context:context().
+by_number(Context, Number) ->
+    E164 = wnm_util:to_e164(Number),
+    AccountId = cb_context:account_id(Context),
+    ViewOptions = [{'key', [AccountId, E164]}, 'include_docs'],
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: Build this as a list of keys based on the accounts URI parameters

File: applications/crossbar/src/modules/cb_port_requests.erl [hunk 3]
```diff
@@ -715,6 +724,67 @@ summary(Context) ->
...
+descendants_by_number(Context, Number) ->
+    ViewOptions = [{'keys', build_keys(Context, Number)}, 'include_docs'],
+    crossbar_doc:load_view(
+        ?ALL_PORT_REQ_NUMBERS
+        ,ViewOptions
+        ,cb_context:set_account_db(Context, ?KZ_PORT_REQUESTS_DB)
+        ,fun normalize_view_results/2
+    ).
+
+-spec build_keys(cb_context:context(), ne_binary()) -> [[ne_binary(), ne_binary()],...] | [].
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: spec should return `[ne_binaries()]`.

Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: And you should probably compile this stuff first! Compiler complained immediately.

File: applications/crossbar/doc/port_requests.md [hunk 4]
```diff
@@ -111,3 +111,109 @@ curl -v -X PUT -H "X-Auth-Token: 6cf321ea39fe960d9b855786a0216064" -H "Content-T
...
+    "status": "success"
+}
+```
+
+## Get port request for account and descendants
+
+### Request
+
+- Verb: `GET`
+- Url: `/accounts/{ACCOUNT_ID}/port_requests/descendants?by_number={NUMBER}`
```
Comment (jamesaimonetti) [inline applications/crossbar/doc/port_requests.md]: I thought we were going to do `/accounts/{ACCOUNT_ID}/descendants/port_requests?by_number={NUMBER}`, similar to how cb_webhooks does it.

File: applications/crossbar/src/modules/cb_port_requests.erl [hunk 5]
```diff
@@ -301,7 +305,10 @@ validate(Context) ->
     validate_port_requests(Context, cb_context:req_verb(Context)).
 
 validate(Context, ?PORT_DESCENDANTS) ->
-    read_descendants(Context);
+    case cb_context:req_value(Context, <<"by_number">>) of
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: This should not exist. It should be handled in `summary/1` and `by_number/2`

File: applications/crossbar/src/modules/cb_port_requests.erl (hunk 2 above)
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_port_requests.erl]: I think this should be defined as:

`ViewOptions = [{'keys', summary_keys(Context, E164)}, 'include_docs'],`

where `summary_keys/2` looks like:

```erlang
summary_keys(Context, E164) ->
    summary_keys(Context, E164, props:get_value(<<"accounts">>, cb_context:req_nouns(Context))).

summary_keys(_Context, E164, [AccountId]) ->
    [[AccountId, E164]];
summary_keys(Context, E164, [AccountId, ?DESCENDANTS]) ->
    DescendantIds = descendant_ids(AccountId),
    [[Id, E164] || Id <- [AccountId | DescendantIds]].

descendant_ids(AccountId) ->
    case couch_mgr:get_results(?WH_ACCOUNTS_DB
                               ,<<"accounts/listing_by_descendants">>
                               ,[{'startkey', [AccountId]}
                                 ,{'endkey', [AccountId, wh_json:new()]}
                                ]
                              )
    of
        {'ok', JObjs} -> [wh_json:get_value([<<"value">>, <<"id">>], JObj) || JObj <- JObjs];
        {'error', _E} -> []
    end.
```

This is untested but hopefully conveys the idea?

**Comment by jamesaimonetti** [comment]
@macpie don't forget about me! :)

File: applications/crossbar/doc/port_requests.md (hunk 4 above)
Comment (jamesaimonetti) [inline applications/crossbar/doc/port_requests.md]: I still am not seeing changes related to this. Has there been discussion with @k-anderson or others on this topic? What's the status here?

Comment (macpie) [inline applications/crossbar/doc/port_requests.md]: I did not see anything in cb_webhooks that could help me.

Comment (jamesaimonetti) [inline applications/crossbar/doc/port_requests.md]: https://github.com/2600hz/kazoo/blob/master/applications/crossbar/src/modules/cb_webhooks.erl#L438-450

It checks the params to the accounts endpoint in the URI to determine what it should do.

Comment (macpie) [inline applications/crossbar/doc/port_requests.md]: Oh got it thank you.

Comment (jamesaimonetti) [inline applications/crossbar/doc/port_requests.md]: Update docs with URI change

**Pull request merged**